pkg/cli/upgradeassistant/cmd/migrate/1170.go

```diff
@@ -138,3 +146,138 @@ func migrateJiraInfo() error {
+						continue
+					}
+					s := &ServiceInfo{
+						ServiceName: service.ServiceName,
+						ProductName: service.ProductName,
+						Deleted:     true,
+					}
+					sharedServiceListInEnvMap[project.ProductName][service.ServiceName] = s
+					sharedServiceInfoList[service.ProductName+"-"+service.ServiceName] = s
+					log.Debugf("migrateSharedService: find service %s-%s only in product env %s", service.ProductName, service.ServiceName, project.Namespace)
+				}
+			}
+		}
+		// update shared service in product env to private
+		if needUpdate {
+			err := mongodb.NewProductColl().Update(project)
+			if err != nil {
+				return errors.Wrap(err, "update shared service in product env")
+			}
+			log.Infof("migrateSharedService: update shared service in product %s success", project.Namespace)
+		}
+	}
+
+	// not found shared service in any env, set all service to private
+	if len(sharedServiceListInEnvMap) == 0 {
+		log.Infof("migrateSharedService: not found shared service in any env")
+		_, err = mongodb.NewServiceColl().UpdateMany(context.Background(),
+			bson.M{}, bson.M{"visibility": "private"}, options.Update().SetUpsert(false))
+		if err != nil {
+			return errors.Wrap(err, "update service visibility")
+		}
+		return nil
+	}
+
+	// get all shared service all revision data
+	sharedServiceAllRevisionData := map[string][]*commonmodels.Service{}
+	for key, service := range sharedServiceInfoList {
+		result, err := mongodb.NewServiceColl().ListServiceAllRevisionsAndStatus(service.ServiceName, service.ProductName)
+		if err != nil {
+			log.Warnf("not found shared service %s-%s", service.ProductName, service.ServiceName)
+			continue
+		}
+		sharedServiceAllRevisionData[key] = result
+	}
+
+	// copy shared service template all revision data to each product
+	for projectName, serviceList := range sharedServiceListInEnvMap {
+		for _, service := range serviceList {
+			if sharedServiceAllRevision, ok := sharedServiceAllRevisionData[service.ProductName+"-"+service.ServiceName]; ok {
+				var updateList []interface{}
+				for _, sharedService := range sharedServiceAllRevision {
+					sharedService.ProductName = projectName
+					sharedService.CreateBy = setting.SystemUser
+					sharedService.CreateTime = time.Now().Unix()
+					if service.Deleted {
+						sharedService.Status = setting.ProductStatusDeleting
+					}
+					updateList = append(updateList, sharedService)
+				}
+				_, err = mongodb.NewServiceColl().InsertMany(context.Background(), updateList)
+				if err != nil {
+					return errors.Wrapf(err, "insert service %s-%s data to %s", service.ProductName, service.ServiceName, projectName)
+				}
+				log.Infof("migrateSharedService: copy service %s-%s data to %s", service.ProductName, service.ServiceName, projectName)
+			} else {
+				log.Warnf("migrateSharedService: not found shared service %s-%s revision data", service.ProductName, service.ServiceName)
+			}
+		}
+	}
+	_, err = mongodb.NewServiceColl().UpdateMany(context.Background(),
```

Comment:
errors may occur when the number of services is pretty large, processing service data in batches may be safer